proposals/0108-transaction-size-fee.md

```diff
@@ -0,0 +1,190 @@
+category: Standard
+type: Core
+status: Draft
+created: 2024-01-18
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Add a new transaction fee based on the serialized size of the transaction.
+
+## Motivation
+
+Currently, fees are based entirely on the time it takes to process the
+transaction in a node:
+
+* the fee per signature covers the sigverify time, before execution
+* priority fees are priced on compute units, which cover the virtual machine
+execution time
+
+A transaction takes up more than time resources, however. The network must also
+broadcast the transaction through Turbine so that all nodes can replay it, which
+consumes bandwidth.
+
+To see this in practice, let's imagine we have two transactions. They take the
+same amount of time to execute, meaning that they have the same number of
+signatures to verify, and they consume the same number of compute units.
+
+If one of those transactions has a larger payload, however, it will use up more
+bandwidth, and potentially reduce the overall transaction throughput of the network.
+
+To quantify the impact of transaction payload size, we ran tests and reported
+the results in [this issue](https://github.com/solana-labs/solana/issues/28063).
+We setup a small private test network, and used `solana-bench-tps` to send large
+numbers of transactions with useless random padding bytes, ranging from 0 to 2206.
+
+The test network removed all QUIC congestion control to better capture total
+throughput.
+
+The inner work in each transaction was exactly the same, performing a CPI into 
+the system program to transfer SOL. This way, we fixed the time dimension of
+transaction process and isolated the impact of transaction size.
+
+The base transaction size is 257 bytes, and the tool sent batches of 50k
+transactions. Here is the table of results, taken from
+[this comment](https://github.com/solana-labs/solana/issues/28063#issuecomment-1727151064):
+
+Without congestion control, there is slight decrease in transactions per second
+with more padding bytes. However, the overall variance is larger than any effect
+from padding bytes. So there may be some negligible effect based on transaction
+size.
+
+This test, however, was only on a small network, so we can't come to any
+conclusions about how larger transactions will impact a real network.
+
+## New Terminology
+
+Transaction size fee: the additional lamports paid by a user, based on the number
+of bytes in the serialized transaction.
+
+## Detailed Design
+
+### Fee Calculation
+
+We propose a solution with a few important features:
+
+* the new fee does not impact current transactions
+* the fee contains a non-linear component to dissuade huge transactions,
+especially while we see the impact of large transactions on a real cluster
+* the fee also includes a linear component for the future, if the effect on a
+real cluster is small
+
+Given `total_transaction_bytes` and the fee parameters
```

Comment:
Assume `total_transaction_bytes` is cheaply available to validators after deserializing?